com.unity.netcode.gameobjects/Tests/Runtime/DeferredMessagingTests.cs

```diff
@@ -0,0 +1,1230 @@
+        {
+        }
+
+        public void OnAfterSendMessage<T>(ulong clientId, ref T message, NetworkDelivery delivery, int messageSizeBytes) where T : INetworkMessage
+        {
+        }
+
+        public void OnBeforeReceiveMessage(ulong senderId, Type messageType, int messageSizeBytes)
+        {
+        }
+
+        public void OnAfterReceiveMessage(ulong senderId, Type messageType, int messageSizeBytes)
+        {
+        }
+
+        public void OnBeforeSendBatch(ulong clientId, int messageCount, int batchSizeInBytes, NetworkDelivery delivery)
+        {
+        }
+
+        public void OnAfterSendBatch(ulong clientId, int messageCount, int batchSizeInBytes, NetworkDelivery delivery)
+        {
+        }
+
+        public void OnBeforeReceiveBatch(ulong senderId, int messageCount, int batchSizeInBytes)
+        {
+        }
+
+        public void OnAfterReceiveBatch(ulong senderId, int messageCount, int batchSizeInBytes)
+        {
+        }
+
+        public bool OnVerifyCanSend(ulong destinationId, Type messageType, NetworkDelivery delivery)
+        {
+            return true;
+        }
+
+        public bool OnVerifyCanReceive(ulong senderId, Type messageType, FastBufferReader messageContent, ref NetworkContext context)
+        {
+            if (messageType == typeof(CreateObjectMessage))
+            {
+                CaughtMessages.Add(new TriggerData
+                {
+                    Reader = new FastBufferReader(messageContent, Allocator.Persistent),
+                    Header = context.Header,
+                    Timestamp = context.Timestamp,
+                    SenderId = context.SenderId,
+                    SerializedHeaderSize = context.SerializedHeaderSize
+                });
+                return false;
+            }
+
+            return true;
+        }
+
+        public void OnBeforeHandleMessage<T>(ref T message, ref NetworkContext context) where T : INetworkMessage
+        {
+        }
+
+        public void OnAfterHandleMessage<T>(ref T message, ref NetworkContext context) where T : INetworkMessage
+        {
+        }
+    }
+
+    public class DeferredMessageTestRpcComponent : NetworkBehaviour
+    {
+        public bool ClientRpcCalled;
+
+        [ClientRpc]
+        public void SendTestClientRpc()
+        {
+            ClientRpcCalled = true;
+        }
+    }
+
+    public class DeferredMessageTestNetworkVariableComponent : NetworkBehaviour
+    {
+        public NetworkVariable<int> TestNetworkVariable = new NetworkVariable<int>();
+    }
+
+    public class DeferredMessageTestRpcAndNetworkVariableComponent : NetworkBehaviour
+    {
+        public bool ClientRpcCalled;
+
+        [ClientRpc]
+        public void SendTestClientRpc()
+        {
+            ClientRpcCalled = true;
+        }
+
+        public NetworkVariable<int> TestNetworkVariable = new NetworkVariable<int>();
+    }
```

Comment:
what was the rationale behind separating these two? could we have one comp with NetVar & RPC in it?

Reply:
Test isolation. If a NetworkVariable is present, NetworkVariableDelta messages will get sent. Makes it hard for a test to reliably verify that the message that was captured was the RPC message.